File: Clinical-code-generator/output/table_descriptive_20260622_120333.R
# ==============================================================================
# PROGRAM NAME  : t_ae_summary.R
# STUDY         : STUDY-001
# PROTOCOL      : D1234C00001
# SPONSOR       : AstraZeneca
# PURPOSE       : Generate descriptive statistics summary
# INPUT         : adsl, adae
# OUTPUT        : t_ae_summary.rtf
# ANALYSIS TYPE : Descriptive
# OUTPUT TYPE   : Table
# POPULATION    : Safety
#
# AUTHOR        : [Programmer Name]
# DATE CREATED  : 22-Jun-2026
# DATE MODIFIED : 22-Jun-2026
#
# REGULATORY    : FDA
# SUBMISSION    : NDA
#
# PACKAGES      : tidyverse, haven, pharmaverse (admiral, rtables, tern)
# VALIDATION    : Independent double programming required
# ==============================================================================

# --- Load Required Packages ---
library(tidyverse)
library(haven)
library(labelled)
library(gtsummary)
library(rtables)
library(tern)
library(r2rtf)
library(flextable)
library(officer)

# --- Program Setup ---
study_id    <- "STUDY-001"
protocol    <- "D1234C00001"
prog_name   <- "t_ae_summary"
output_name <- "t_ae_summary"
pop_flag    <- "SAFFL"
trt_var     <- "TRTA"
alpha       <- 0.05
conf_level  <- 0.95

# --- Path Configuration ---
adam_path   <- "/data/adam"
output_path <- "/output/tables"

# --- Treatment Order ---
trt_levels <- c("Placebo", "Treatment A", "Treatment B", "Total")
trt_n_labels <- c()  # Will be populated after data read

# --- Read Input Datasets ---
# Read ADSL
adsl <- read_sas(file.path(adam_path, "adsl.sas7bdat")) %>%
    filter(SAFFL == "Y") %>%
    mutate(
        TRTA = factor(
            TRTA,
            levels = trt_levels
        )
    )

cat(sprintf("NOTE: ADSL has %d records after filtering.\n", nrow(adsl)))
# Read ADAE
adae <- read_sas(file.path(adam_path, "adae.sas7bdat")) %>%
    filter(SAFFL == "Y") %>%
    mutate(
        TRTA = factor(
            TRTA,
            levels = trt_levels
        )
    )

cat(sprintf("NOTE: ADAE has %d records after filtering.\n", nrow(adae)))

# --- Get Population Counts (Big N) ---
big_n <- adsl %>%
    distinct(USUBJID, .keep_all = TRUE) %>%
    count(TRTA, name = "N") %>%
    mutate(
        N_label = sprintf("%s\n(N=%d)", TRTA, N)
    )

cat("NOTE: Population counts:\n")
print(big_n)

# --- Data Preparation ---
analysis_data <- adsl %>%
    # Add treatment numeric ordering
    mutate(
        TRTN = case_when(
            TRTA == "Placebo"      ~ 1L,
            TRTA == "Treatment A"  ~ 2L,
            TRTA == "Treatment B"  ~ 3L,
            TRUE                              ~ 99L
        )
    ) %>%
    arrange(TRTN, USUBJID, AVISITN)

# --- Create Total group ---
analysis_with_total <- bind_rows(
    analysis_data,
    analysis_data %>% mutate(TRTA = "Total", TRTN = 99L)
)

cat(sprintf("NOTE: Analysis dataset has %d records (including Total).\n", 
    nrow(analysis_with_total)))

# --- Descriptive Statistics ---
desc_stats <- analysis_with_total %>%
    group_by(TRTA, TRTN) %>%
    summarise(
        across(
            c(AVAL, CHG),
            list(
                n      = ~sum(!is.na(.)),
                mean   = ~mean(., na.rm = TRUE),
                sd     = ~sd(., na.rm = TRUE),
                median = ~median(., na.rm = TRUE),
                min    = ~min(., na.rm = TRUE),
                max    = ~max(., na.rm = TRUE),
                q1     = ~quantile(., 0.25, na.rm = TRUE),
                q3     = ~quantile(., 0.75, na.rm = TRUE)
            ),
            .names = "{.col}_{.fn}"
        ),
        .groups = "drop"
    ) %>%
    arrange(TRTN)

# --- Format for display ---
desc_formatted <- desc_stats %>%
    mutate(
        n_c      = sprintf("%d", AVAL_n),
        mean_sd  = sprintf("%.1f (%.2f)", AVAL_mean, AVAL_sd),
        median_c = sprintf("%.1f", AVAL_median),
        q1_q3    = sprintf("%.1f, %.1f", AVAL_q1, AVAL_q3),
        min_max  = sprintf("%.1f, %.1f", AVAL_min, AVAL_max)
    ) %>%
    select(TRTA, TRTN, n_c, mean_sd, median_c, q1_q3, min_max)

# --- Using rtables for submission-ready output ---
lyt <- basic_table() %>%
    split_cols_by("TRTA") %>%
    add_colcounts() %>%
    analyze(
        vars = c("AVAL", "CHG"),
        afun = function(x) {
            in_rows(
                "n"           = rcell(sum(!is.na(x)), format = "xx"),
                "Mean (SD)"   = rcell(c(mean(x, na.rm=TRUE), sd(x, na.rm=TRUE)), format = "xx.x (xx.xx)"),
                "Median"      = rcell(median(x, na.rm=TRUE), format = "xx.x"),
                "Q1, Q3"      = rcell(c(quantile(x, 0.25, na.rm=TRUE), quantile(x, 0.75, na.rm=TRUE)), format = "xx.x, xx.x"),
                "Min, Max"    = rcell(c(min(x, na.rm=TRUE), max(x, na.rm=TRUE)), format = "xx.x, xx.x")
            )
        }
    )

result_table <- build_table(lyt, analysis_with_total)
print(result_table)

# --- Generate Table Output ---

# Using r2rtf for submission-ready RTF output
output_file <- file.path(output_path, paste0(output_name, ".rtf"))

result_df %>%
    r2rtf::rtf_title(
        title = "Table X.X.X",
        subtitle = "Generate descriptive statistics summary"
    ) %>%
    r2rtf::rtf_colheader(
        colheader = "Statistic | Placebo\n(N={N1}) | Treatment A\n(N={N2}) | Treatment B\n(N={N3})",
        col_rel_width = c(3, 2, 2, 2)
    ) %>%
    r2rtf::rtf_body(
        col_rel_width = c(3, 2, 2, 2),
        text_justification = c("l", "c", "c", "c")
    ) %>%
    r2rtf::rtf_footnote(
        footnote = c(
            "Source: adsl, adae",
            "Program: {prog_name}.R  Output: {output_name}.rtf",
            paste0("Generated: ", format(Sys.time(), "%d%b%Y %H:%M"))
        )
    ) %>%
    r2rtf::rtf_encode() %>%
    r2rtf::write_rtf(output_file)

cat(sprintf("NOTE: Output written to %s\n", output_file))

# --- Validation Checks ---

# Check for missing critical variables
n_missing_usubjid <- sum(is.na(analysis_data$USUBJID))
n_missing_trt <- sum(is.na(analysis_data$TRTA))

cat(sprintf("\nValidation Results:\n"))
cat(sprintf("  Missing USUBJID: %d\n", n_missing_usubjid))
cat(sprintf("  Missing Treatment: %d\n", n_missing_trt))
cat(sprintf("  Total subjects: %d\n", n_distinct(analysis_data$USUBJID)))

# Session info for reproducibility
cat("\n--- Session Info ---\n")
sessionInfo()

cat("\n========================================\n")
cat(sprintf("Program: %s.R\n", prog_name))
cat(sprintf("Output:  %s.rtf\n", output_name))
cat("Status:  COMPLETE\n")
cat("========================================\n")

File: Clinical-code-generator/output/table_descriptive_20260622_120333.py
"""
================================================================================
PROGRAM NAME  : t_ae_summary.py
STUDY         : STUDY-001
PROTOCOL      : D1234C00001
SPONSOR       : AstraZeneca
PURPOSE       : Generate descriptive statistics summary
INPUT         : adsl, adae
OUTPUT        : t_ae_summary.rtf
ANALYSIS TYPE : Descriptive
OUTPUT TYPE   : Table
POPULATION    : Safety

AUTHOR        : [Programmer Name]
DATE CREATED  : 22-Jun-2026
DATE MODIFIED : 22-Jun-2026

REGULATORY    : FDA
SUBMISSION    : NDA

VALIDATION    : Independent double programming required
CDISC VERSION : ADaM IG v1.3 / SDTM IG v3.4
================================================================================
"""

import pandas as pd
import numpy as np
from pathlib import Path
from datetime import datetime
import warnings
warnings.filterwarnings("ignore")
from scipy import stats
import statsmodels.api as sm
from openpyxl import Workbook
from docx import Document

# --- Program Configuration ---
CONFIG = {
    'study_id': 'STUDY-001',
    'protocol': 'D1234C00001',
    'prog_name': 't_ae_summary',
    'output_name': 't_ae_summary',
    'pop_flag': 'SAFFL',
    'trt_var': 'TRTA',
    'alpha': 0.05,
    'conf_level': 0.95,
    'adam_path': Path(r'/data/adam'),
    'output_path': Path(r'/output/tables'),
}

# Treatment ordering
TRT_ORDER = ['Placebo', 'Treatment A', 'Treatment B', 'Total']
TRT_MAP = {'Placebo': 1, 'Treatment A': 2, 'Treatment B': 3, 'Total': 99}

# Ensure output directory exists
CONFIG['output_path'].mkdir(parents=True, exist_ok=True)

# --- Read Input Datasets ---
import pyreadstat

# Read ADSL
adsl_path = CONFIG['adam_path'] / 'adsl.sas7bdat'
adsl_raw, meta = pyreadstat.read_sas7bdat(str(adsl_path))
adsl = adsl_raw[(df["SAFFL"] == "Y")].copy()
adsl[CONFIG['trt_var']] = pd.Categorical(
    adsl[CONFIG['trt_var']], categories=TRT_ORDER[:-1], ordered=True
)
print(f"NOTE: ADSL has {adsl.shape[0]} records after filtering.")
# Read ADAE
adae_path = CONFIG['adam_path'] / 'adae.sas7bdat'
adae_raw, meta = pyreadstat.read_sas7bdat(str(adae_path))
adae = adae_raw[(df["SAFFL"] == "Y")].copy()
adae[CONFIG['trt_var']] = pd.Categorical(
    adae[CONFIG['trt_var']], categories=TRT_ORDER[:-1], ordered=True
)
print(f"NOTE: ADAE has {adae.shape[0]} records after filtering.")

# --- Get Population Counts (Big N) ---
big_n = adsl.drop_duplicates(subset=['USUBJID']).groupby(
    CONFIG['trt_var']
)['USUBJID'].nunique().to_dict()

print(f"NOTE: Population counts: {big_n}")

# --- Data Preparation ---
analysis_data = adsl.copy()

# Add treatment numeric ordering
analysis_data['TRTN'] = analysis_data[CONFIG['trt_var']].map(TRT_MAP)

# Sort data
analysis_data = analysis_data.sort_values(
    ['TRTN'] + ['USUBJID', 'AVISITN']
).reset_index(drop=True)

# Create version with Total group
analysis_total = analysis_data.copy()
analysis_total[CONFIG['trt_var']] = 'Total'
analysis_total['TRTN'] = 99

analysis_with_total = pd.concat(
    [analysis_data, analysis_total], ignore_index=True
)

print(f"NOTE: Analysis dataset has {analysis_with_total.shape[0]} records (including Total).")

# --- Descriptive Statistics ---

def compute_descriptive_stats(df, var, group_var):
    """Compute descriptive statistics by treatment group."""
    stats_df = df.groupby(group_var)[var].agg([
        ('n', 'count'),
        ('mean', 'mean'),
        ('std', 'std'),
        ('median', 'median'),
        ('min', 'min'),
        ('max', 'max'),
        ('q1', lambda x: x.quantile(0.25)),
        ('q3', lambda x: x.quantile(0.75))
    ]).reset_index()
    
    # Format for display
    stats_df['n_c'] = stats_df['n'].apply(lambda x: f"{x:.0f}")
    stats_df['mean_sd'] = stats_df.apply(
        lambda r: f"{r['mean']:.1f} ({r['std']:.2f})", axis=1
    )
    stats_df['median_c'] = stats_df['median'].apply(lambda x: f"{x:.1f}")
    stats_df['q1_q3'] = stats_df.apply(
        lambda r: f"{r['q1']:.1f}, {r['q3']:.1f}", axis=1
    )
    stats_df['min_max'] = stats_df.apply(
        lambda r: f"{r['min']:.1f}, {r['max']:.1f}", axis=1
    )
    
    return stats_df

# Compute statistics for each analysis variable
results = {}
for var in ['AVAL', 'CHG']:
    results[var] = compute_descriptive_stats(
        analysis_with_total, var, CONFIG['trt_var']
    )
    print(f"\nDescriptive Statistics for {var}:")
    print(results[var][['TRTA', 'n_c', 'mean_sd', 'median_c', 'q1_q3', 'min_max']])

# --- Generate Table Output ---

output_file = CONFIG['output_path'] / f"{CONFIG['output_name']}.rtf"

# Create formatted output table
from openpyxl import Workbook
from openpyxl.styles import Font, Alignment, Border, Side

wb = Workbook()
ws = wb.active
ws.title = "Results"

# Header information
ws['A1'] = CONFIG['study_id']
ws['A1'].font = Font(bold=True, size=12)
ws['A2'] = f"Protocol: {CONFIG['protocol']}"
ws['A3'] = f"Table X.X.X: Generate descriptive statistics summary"
ws['A3'].font = Font(bold=True, size=11)
ws['A4'] = f"Population: Safety Population"

# Column headers
headers = ['Statistic', 'Placebo\n(N={})', 'Treatment A\n(N={})', 'Treatment B\n(N={})']
for col, header in enumerate(headers, 1):
    cell = ws.cell(row=6, column=col, value=header)
    cell.font = Font(bold=True)
    cell.alignment = Alignment(horizontal='center', wrap_text=True)

# Write results data
# (Implementation would write actual results from analysis)

# Footnotes
last_row = ws.max_row + 2
ws.cell(row=last_row, column=1, value=f"Source: adsl, adae")
ws.cell(row=last_row + 1, column=1, value=f"Program: {CONFIG['prog_name']}.py")
ws.cell(row=last_row + 2, column=1, value=f"Generated: {datetime.now().strftime('%d%b%Y %H:%M')}")

wb.save(str(output_file))
print(f"NOTE: Table written to {output_file}")

# --- Validation Checks ---

print("\n" + "=" * 60)
print("VALIDATION RESULTS")
print("=" * 60)

# Check for missing critical variables
n_missing_usubjid = analysis_data['USUBJID'].isna().sum()
n_missing_trt = analysis_data[CONFIG['trt_var']].isna().sum()
n_subjects = analysis_data['USUBJID'].nunique()

print(f"  Missing USUBJID: {n_missing_usubjid}")
print(f"  Missing Treatment: {n_missing_trt}")
print(f"  Total unique subjects: {n_subjects}")
print(f"  Total records: {analysis_data.shape[0]}")

# Verify population counts
assert n_missing_usubjid == 0, "ERROR: Missing USUBJID values found!"
assert n_missing_trt == 0, "ERROR: Missing treatment values found!"

print("\n" + "=" * 60)
print(f"Program: {CONFIG['prog_name']}.py")
print(f"Output:  {CONFIG['output_name']}.rtf")
print(f"Status:  COMPLETE")
print(f"Time:    {datetime.now().strftime('%d%b%Y %H:%M:%S')}")
print("=" * 60)